File: javascript/codeql/request_forgery/benign_edge.js
const axios = require("axios");
const express = require("express");

const app = express();
const paths = new Map([
  ["health", "/health"],
  ["status", "/status"],
]);

app.get("/fetch-same-origin", async (req, res) => {
  const path = paths.get(String(req.query.path || "health")) || paths.get("health");
  const url = new URL(path, "https://api.example.invalid");
  const response = await axios.get(url.toString());
  res.send(response.data);
});

File: python/codeql/full_ssrf/benign_edge.py
from urllib.parse import urlparse

import requests
from flask import Flask, request


app = Flask(__name__)


@app.get("/fetch-same-origin")
def fetch_same_origin():
    url = request.args.get("url", "https://api.example.invalid/health")
    parsed = urlparse(url)
    if parsed.scheme != "https" or parsed.hostname != "api.example.invalid":
        return "blocked", 400
    return requests.get(url, timeout=5).text
